sql/core/src/test/scala/org/apache/spark/sql/MeltSuite.scala

```diff
@@ -0,0 +1,297 @@
+    )
+  }
+
+  test("melt with invalid arguments") {
+    // melting with column in both ids and values
+    assertException[IllegalArgumentException] {
+      Melt.of(meltWideDataDs, Seq("str1", "int1"), Seq("str1", "str2", "int1", "long1"))
+    }("A column cannot be both an id and a value column: str1, int1")
+
+    // melting with empty list of value columns
+    // where potential value columns are of incompatible types
+    assertException[IllegalArgumentException] {
+      Melt.of(meltWideDataDs, Seq.empty, Seq.empty)
+    }("All values must be of compatible types, " +
+      "types StringType and LongType are not compatible")
+    assertException[IllegalArgumentException] {
+      Melt.of(meltWideDataDs, Seq("id"), Seq.empty)
+    }("All values must be of compatible types, " +
+      "types StringType and IntegerType are not compatible")
+
+    // melting without giving values and no non-id columns
+    assertException[IllegalArgumentException] {
+      Melt.of(meltWideDataDs.select("id"), Seq("id"))
+    }("The dataset has no non-id columns to melt")
+
+    // melting with id column `variable`
+    assertException[IllegalArgumentException] {
+      Melt.of(
+        meltWideDataDs.withColumnRenamed("id", "variable"),
+        Seq("variable"), Seq("str1", "str2")
+      )
+    }("Column name for variable column (variable) must not exist among id columns: variable")
+    checkAnswer(
+      Melt.of(meltWideDataDs.withColumnRenamed("id", "variable"),
+        Seq("variable"), Seq("str1", "str2"), variableColumnName = "var", valueColumnName = "val"),
+      meltedWideDataRows
+    )
+
+    // melting with id column `value`
+    assertException[IllegalArgumentException] {
+      Melt.of(meltWideDataDs.withColumnRenamed("id", "value"),
+        Seq("value"), Seq("str1", "str2"))
+    }("Column name for value column (value) must not exist among id columns: value")
+    checkAnswer(
+      Melt.of(meltWideDataDs.withColumnRenamed("id", "value"),
+        Seq("value"), Seq("str1", "str2"), variableColumnName = "var", valueColumnName = "val"),
+      meltedWideDataRows
+    )
+
+  }
+
+  test("melt with dot and backtick") {
+    val df = meltWideDataDs
+      .withColumnRenamed("id", "an.id")
+      .withColumnRenamed("str1", "str.one")
+      .withColumnRenamed("str2", "str.two")
+    checkAnswer(
+      Melt.of(df, Seq("`an.id`"), Seq("`str.one`", "`str.two`")),
+      meltedWideDataRows.map(row => Row(
+        row.getInt(0),
+        row.getString(1) match {
+          case "str1" => "str.one"
+          case "str2" => "str.two"
+        },
+        row.getString(2)
+      ))
+    )
+
+    // without backticks, this references struct fields, which do not exist
+    assertException[IllegalArgumentException] {
+      Melt.of(df, Seq("an.id"), Seq("str.one", "str.two")).collect()
+    }("Unknown columns: an.id, str.one, str.two, dataset has: an.id, str.one, str.two, int1, long1")
+  }
+
+  /** TODO: Would be nice to melt on struct fields.
```

Comment:
Let's file a separate JIRA for this, and make it to `TODO(SPARK-XXXX): ... `

Reply:
Done in 080a736e17